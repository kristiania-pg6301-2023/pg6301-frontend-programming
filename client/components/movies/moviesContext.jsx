import React from "react";

export const MoviesContext = React.createContext({
  fetchMovies: async () => [],
  onAddMovie: async (movie) => {},
});
