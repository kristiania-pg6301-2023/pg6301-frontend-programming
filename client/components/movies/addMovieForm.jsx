import React, { useContext, useState } from "react";
import { MoviesContext } from "./moviesContext";

export function AddMovieForm() {
  const [title, setTitle] = useState("");
  const { onAddMovie } = useContext(MoviesContext);

  async function handleSubmit(e) {
    e.preventDefault();

    const movie = { title };
    await onAddMovie(movie);
    // reload
  }

  return (
    <form onSubmit={handleSubmit}>
      <h2>Add movie</h2>
      <div>
        <input value={title} onChange={(e) => setTitle(e.target.value)} />
      </div>
      <div>
        <button>Submit</button>
      </div>
    </form>
  );
}
