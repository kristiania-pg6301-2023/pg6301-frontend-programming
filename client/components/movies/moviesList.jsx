import React, { useContext, useEffect, useState } from "react";
import { MoviesContext } from "./moviesContext";

export function MoviesList() {
  const [movies, setMovies] = useState([]);
  const [loading, setLoading] = useState(true);

  const { fetchMovies } = useContext(MoviesContext);

  async function loadMovies() {
    setLoading(true);
    setMovies(await fetchMovies());
    setLoading(false);
  }

  useEffect(() => {
    loadMovies();
  }, []);

  return (
    <>
      <h2>List the movies</h2>
      {loading && <div>Spinner</div>}
      {movies.map((m) => (
        <div key={m._id}>{m.title}</div>
      ))}
    </>
  );
}
