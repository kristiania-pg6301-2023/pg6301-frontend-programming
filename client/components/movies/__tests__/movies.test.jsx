import renderer, { act } from "react-test-renderer";
import { MoviesList } from "../moviesList";
import { MoviesContext } from "../moviesContext";
import { AddMovieForm } from "../addMovieForm";

describe("movies react application", () => {
  it("show loading screen", () => {
    const component = renderer.create(<MoviesList />);
    expect(component).toMatchSnapshot();
  });

  it("show loaded movies", async () => {
    const fetchMovies = () => [
      { title: "Jest - the movie", _id: 1 },
      { title: "Promises - as promised", _id: 2 },
    ];

    let component;
    await act(async () => {
      component = renderer.create(
        <MoviesContext.Provider value={{ fetchMovies }}>
          <MoviesList />
        </MoviesContext.Provider>,
      );
    });
    expect(component).toMatchSnapshot();
  });

  it("add a new movie", async () => {
    const onAddMovie = jest.fn();
    const component = renderer.create(
      <MoviesContext.Provider value={{ onAddMovie }}>
        <AddMovieForm />
      </MoviesContext.Provider>,
    );

    await act(async () => {
      component.root.findByType("input").props.onChange({
        target: { value: "My test movie" },
      });
    });
    await act(async () => {
      component.root.findByType("form").props.onSubmit({
        preventDefault() {},
      });
    });

    expect(onAddMovie).toBeCalledWith({
      title: "My test movie",
    });
  });
});
