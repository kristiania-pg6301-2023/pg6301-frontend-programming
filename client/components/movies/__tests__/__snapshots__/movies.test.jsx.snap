// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`movies react application show loaded movies 1`] = `
[
  <h2>
    List the movies
  </h2>,
  <div>
    Jest - the movie
  </div>,
  <div>
    Promises - as promised
  </div>,
]
`;

exports[`movies react application show loading screen 1`] = `
[
  <h2>
    List the movies
  </h2>,
  <div>
    Spinner
  </div>,
]
`;
